util/src/matrix.rs

```diff
@@ -0,0 +1,541 @@
+use std::iter::{DoubleEndedIterator, ExactSizeIterator, FusedIterator, Iterator};
+use std::marker::PhantomData;
+use std::mem::{ManuallyDrop, MaybeUninit};
+use std::ops::{Index, IndexMut};
+use std::ptr;
+use std::slice;
+
+/// Calculates the number of elements in a matrix with the given height and width, panicking on
+/// overflow.
+fn size_checked(height: usize, width: usize) -> usize {
+    height.checked_mul(width).expect("memory overflow")
+}
+
+/// Returns an uninitialized `Vec` of lenth `len`.
+fn uninit_vec<T>(len: usize) -> Vec<MaybeUninit<T>> {
+    let mut res: Vec<MaybeUninit<T>> = Vec::with_capacity(len);
+    unsafe {
+        // SAFETY: `with_capacity` guarantees that `res.capacity() == len`, so we're not violating
+        // memory rules. `MaybeUninit` doesn't need initialization, so the values are already valid.
+        debug_assert_eq!(res.capacity(), len);
+        res.set_len(len);
+    }
+    res
+}
+
+/// A non-resizeable matrix.
+pub struct Matrix<T> {
+    data: *mut T,
```

Comment:
I think you mentioned wanting to avoid `Vec` since we don't need resizing. How about `Box<[T]>`? Seems like that could reduce unsafe code while avoiding the two fields of `Vec`.